sw/device/lib/crypto/impl/aes.c

```diff
@@ -0,0 +1,267 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/crypto/include/aes.h"
+
+#include "sw/device/lib/base/memory.h"
+#include "sw/device/lib/crypto/drivers/aes.h"
+#include "sw/device/lib/crypto/impl/integrity_check.h"
+#include "sw/device/lib/crypto/include/datatypes.h"
+
+/**
+ * Extract an AES key from the blinded key struct.
+ *
+ * @param blinded_key Blinded key struct.
+ * @param aes_mode Block cipher mode.
+ * @param[out] aes_key Destination AES key struct.
+ * @return Result of the operation.
+ */
+static crypto_status_t construct_aes_key(
+    const crypto_blinded_key_t *blinded_key, const block_cipher_mode_t aes_mode,
+    aes_key_t *aes_key) {
+  // TODO(#15590): add support for sideloaded keys by actuating keymgr here if
+  // needed (this requires a keymgr driver).
+  if (blinded_key->config.hw_backed != kHardenedBoolFalse) {
+    return kCryptoStatusBadArgs;
+  }
+  aes_key->sideload = kHardenedBoolFalse;
+
+  // Interpret the key mode and check that it matches the requested block
+  // cipher mode.
+  switch (blinded_key->config.key_mode) {
+    case kKeyModeAesEcb:
+      if (aes_mode != kBlockCipherModeEcb) {
+        return kCryptoStatusBadArgs;
+      }
+      aes_key->mode = kAesCipherModeEcb;
```

Comment:
This is something I've been wondering about. There's a lot of code size (and potential control-flow vulnerability points) here to support different enums that mean the same thing. I could update the AES driver to accept the same type for this enum as the higher-level API, but that would mean importing to the AES driver some header that contains that type (currently it's in `include/aes.h`). Currently the drivers are somewhat standalone -- API endpoints depend on drivers but not the other way around -- and this would break that property. But maybe it's for the best, because I think pretty much every algorithm will have this duplicate-enum issue otherwise. WDYT?

Reply:
I thought about it a little more and made the following adjustments:
- The mode is now represented using the same high-HW values in both the top-level API and the driver, but rather than having one import the other (which is a little circular) I include a static assert in `impl/aes.c`, which imports them both, to check that the values match. Converting from the high-HW values to the hardware encoding is now done inside the driver.
- The key length enum is now replaced with the actual word-length of the key (the API didn't have a high-HW enum for this), which simplifies and shortens the code. Again, converting to the hardware encoding is now in the driver.